### DISCUSSION TITLE
Show more support for FOSS projects

### DISCUSSION THREAD

**Discussion by solonovamax** [modrinth/.github#54] [Feature requests]
**Is your feature request related to a problem? Please describe.**
Many people, like myself, prefer to support FOSS mods.
Here's a few ideas I had to help promote FOSS mods, because open source is cool!

**Describe the solution you'd like**
Here's just a random dump of ideas that I think are cool/should be mentioned:
- When searching for mods, allow users to filter by "Only FOSS mods". This shows mods of _any_ FOSS license, and doesn't show any closed source mods. Currently, you can filter by a specific FOSS license, but I don't think you can filter by *all* licenses.
- When searching for mods regularly, there is a badge beside each mod indicating whether or not it is FOSS.
  - A good example of this is the site [AlternativeTo](https://alternativeto.net/), which shows either a "Freemium • Proprietary", "Freemium • Open Source", "Paid • Proprietary", "Free • Proprietary", or a "Free • Open Source" tag beside each item.
  - Though, instead of putting a text element on it, you can always do something like adding [an icon](https://icons8.com/icon/63655/open-source) (<img src="https://img.icons8.com/color/452/open-source--v1.png" height=16 style="display: inline-block;">) beside the mod to indicate that it is an open source project.
- FOSS projects rank better in search results. Okay, I know this isn't the right repo for this, but I just want to mention it. Honestly, I'm not sure whether I like or dislike that idea, but I just wanna put it out there to put it on peoples' radars.
- "Contributors" integration with github. Not that hard to do, and it would look pretty nice to have in the sidebar. Just use the GH api to fetch the 4-5 top contributors for each (FOSS) project, and then displat 'em on the side.

**Describe alternatives you've considered**
Alternatively, you could just *not* do it. There's nothing forcing you too, tbh.
But I would love to see the fabric open source modding community thrive, and I think this would definitely help it. For now, the majority of mods on Modrinth seem to be FOSS, but if Modrinth because mainstream like Curseforge did, that could easily change.

**Comment by triphora** [comment]
Here is my personal feedback for each of your ideas.

> When searching for mods, allow users to filter by "Only FOSS mods". This shows mods of any FOSS license, and doesn't show any closed source mods. Currently, you can filter by a specific FOSS license, but I don't think you can filter by all licenses.

imo, the best implementation for this would be to allow searches for multiple licenses at once (would require a backend change). The number of people looking for FOSS-only mods would be relatively low, I would think, since FOSS diehards wouldn't be playing Minecraft, but I can see the benefits to multi-license search.

> When searching for mods regularly, there is a badge beside each mod indicating whether or not it is FOSS.

While Modrinth wouldn't be able to use that logo you provided specifically ([see here](https://opensource.org/trademark-guidelines#Business_web_sites); well I mean theoretically it could, but it's still a bit iffy), I don't think that's a bad idea. This would get a bit tricky with custom licenses, though.

> FOSS projects rank better in search results.

This is the only idea I'm completely opposed to. It does seem a bit cheaty and would introduce a huge bias against a large portion of Forge mods.

> "Contributors" integration with github

Not sure how to feel on this one. It's hit or miss even on GH, and would introduce inconsistencies from mod page to mod page. Plus, it'd require an API call to GH in addition to the calls being made to labrinth, which could slow things down quite a bit (even if cached).

**Comment by Minenash** [comment]
> When searching for mods, allow users to filter by "Only FOSS mods". This shows mods of any FOSS license, and doesn't show any closed source mods. Currently, you can filter by a specific FOSS license, but I don't think you can filter by all licenses.

One way to do this might be to add a certain "license" to the drop down: `All FOSS Licenses`. I disagree with waffles, because I feel like there would be more people want to look at any FOSS mod opposed to only a very specific license, like it is rn.

> When searching for mods regularly, there is a badge beside each mod indicating whether or not it is FOSS

I like this one, I wouldn't have an icon if it isn't FOSS, but a nice icon to signify it's FOSS would be pretty nice.

> FOSS projects rank better in search results

No.  As much as I would want this, I think it would be a terrible decision.

> "Contributors" integration with github. Not that hard to do, and it would look pretty nice to have in the sidebar. Just use the GH api to fetch the 4-5 top contributors for each (FOSS) project, and then display 'em on the side.

I'm opposed to what you exactly wrote for the reasons waffles stated, but it would be kinda cool to be able to add "team members" that aren't on modrinth and connect them with a url (github, personal, ect.) because it would be nice to give credit to authors/contributors, even if they don't have a modrinth account.

**Comment by solonovamax** [comment]
> imo, the best implementation for this would be to allow searches for multiple licenses at once (would require a backend change). The number of people looking for FOSS-only mods would be relatively low, I would think, since FOSS diehards wouldn't be playing Minecraft, but I can see the benefits to multi-license search.

I mean, why not make it the default setting? And I'm kinda a FOSS-diehard tbh.
But just because there won't be many "foss diehards", doesn't mean you can't push FOSS to new users.

> While Modrinth wouldn't be able to use that logo you provided specifically (see here; well I mean theoretically it could, but it's still a bit iffy), I don't think that's a bad idea. This would get a bit tricky with custom licenses, though.

yeah, I didn't bother spending a lot of time to find an icon that ""represents open source"", because I'm lazy as shit. I just linked the first thing I found lol

As for custom licenses, just say "if it's not a super well-known license (ei. one that's already in the system), then you don't get the badge. Pretty simple.

> This is the only idea I'm completely opposed to. It does seem a bit cheaty and would introduce a huge bias against a large portion of Forge mods.

yeah,,, I'm not really sure where I stand either with that. Because on one hand it does sound kind of good because it promotes FOSS, but on the other it also has a bias towards FOSS mods. I'm kind of fine with having a bias towards FOSS mods, but don't like it at the same time? I just put it there to throw it out.

> Not sure how to feel on this one. It's hit or miss even on GH, and would introduce inconsistencies from mod page to mod page. Plus, it'd require an API call to GH in addition to the calls being made to labrinth, which could slow things down quite a bit (even if cached).

That's very fair. Though, cacheing the value and only re-requesting it once per day would make the load a lot less, but if you end up with like hundreds of thousands of mods, then it might not scale up to that. On the other hand, there are 86,400 seconds in a day, so if you make at most 1 request a second handled on a separate thread from the main process, it should scale perfectly fine up to 86k mods. The github api ratelimit is 5,000 requests per hour. So that's 120,000 requests per day. Once modrinth gets more than 120k mods on it, you could just change it from updating once per day to updating once per 2 days. If this is implemented that is.

---

> One way to do this might be to add a certain "license" to the drop down: All FOSS Licenses. I disagree with waffles, because I feel like there would be more people want to look at any FOSS mod opposed to only a very specific license, like it is rn.

pretty much this

> I like this one, I wouldn't have an icon if it isn't FOSS, but a nice icon to signify it's FOSS would be pretty nice.

Yeah, the only question is "what icon?", since other than the OSI icon, I'm not sure if there's a "common" icon for "open source". I guess you could use like a yellow star or smth? But that just seems kinda lame tbh. The github logo is always an option (if it's allowed under their license), but then that kind of excludes gitlab/bitbucket/etc.

> I'm opposed to what you exactly wrote for the reasons waffles stated, but it would be kinda cool to be able to add "team members" that aren't on modrinth and connect them with a url (github, personal, ect.) because it would be nice to give credit to authors/contributors, even if they don't have a modrinth account.

That's very fair. Though, the aspect I was targeting was more "add contributors automatically" instead of "add contributors manually", but that works too. I just thought it would be nice to have an automatic list of contributors.
Also, "repository health" stuff could be nice, like some graphs or smth. Idk.
Really, the whole interfacing with the github api thing could be moved from the server to the client tbh. Rather than having the server update all that, just make the client request it from github if the page needs it, then embed it into the page.

**Comment by PssbleTrngle** [comment]
I think giving the user the ability to browse for only FOSS projects would be great, but I also think it could be implemented by an exclude filter in an advanced search page.

Showing a special icon for FOSS-Projects would also be nice, maybe something "community"-related, like a group of people, if there is no better solution.

Like the previous replies, I also strongly disapprove of giving FOSS-Mods higher search priority for reasons already mentioned.

GitHub integration could be nice to show information like "last commit" or even the small commit graph seen in the repository tab if a user, but the actual use case where I need that information is rather rare, only if a mod has not updated for a long time, but I want to check if it is at least worked on.
Most of the time the information about the last released version suffices.

Adding external contributors that don't have an modrinth account would be nice and sounds reasonable to implement, on the other hand most of these will be GitHub users which could simply create a modrinth account by just once logging into modrinth with their GitHub account.

**Comment by Prospector** [comment]
The open source filter exists on the site now.

I'm not sure if any of these other ideas will be implemented. I definitely don't think we will be ranking open source projects higher in search results, and I think the contributors idea is a bit out of scope for Modrinth, that info already exists on the repositories, which is where people interested in the source should go.

The last idea then is the open source label. I don't think making it appear in search would make sense because you can always use the filter, but maybe on project pages we could have "Visible source" and "Open source" labels on projects with sources linked and projects with open source licenses, respectively.
